### PR TITLE
16768 kernel printf should know about %j and %z size specifiers (gcc10)

### PR DESCRIPTION
**Pull request by citrus-it**

### CODE REVIEW

File: gcc/testsuite/gcc.dg/format/cmn-err-1.c [hunk 1]
```diff
@@ -28,6 +30,8 @@ int main()
   cmn_err_func (0, "%hd %hD %ho %hO %hx %hX %hu", h, h, h, h, h, h, h);
   cmn_err_func (0, "%hhd %hhD %hho %hhO %hhx %hhX %hhu",
 		hh, hh, hh, hh, hh, hh, hh);
+  cmn_err_func (0, "%jd %jD %jo %jO %jx %jX %ju", j, j, j, j, j, j, j);
```
Comment (richlowe) [inline gcc/testsuite/gcc.dg/format/cmn-err-1.c:33]: This isn't the kind of test that compares output, and needs a matching change to the stock output, right?

Comment (citrus-it) [reply gcc/testsuite/gcc.dg/format/cmn-err-1.c:33]: It is not. In fact the compiler doesn't know what `cmn_err` does, it just knows that it takes a format string, and knows the rules for format characters. With this test I'm just checking for successful compilation with no warnings. If I run that test on the current version of gcc10, I get a slew of output like:

```
1306681:/home/andy/helios/tmp/omnios-build/gcc-10.4.0/gcc-10.4.0/gcc/testsuite/gcc.dg/format/cmn-err-1.c:33:22: warning: unknown conversion type character 'j' in format [-Wformat=]
1306689:/home/andy/helios/tmp/omnios-build/gcc-10.4.0/gcc-10.4.0/gcc/testsuite/gcc.dg/format/cmn-err-1.c:34:22: warning: unknown conversion type character 'z' in format [-Wformat=]
...
1306679:FAIL: gcc.dg/format/cmn-err-1.c  (test for excess errors)
```

and it passes cleanly with a patched gcc.

I could have also added some negative tests here such as verifying that something like `%zs` produces the expected warning, but I have not. I'm not sure they are that valuable given the change.

Comment (richlowe) [reply gcc/testsuite/gcc.dg/format/cmn-err-1.c:33]: Sounds good.  I didn't remember whether `cmn_err_func` existed, or was just set with the format attribute.  Thanks.

**Comment by rmustacc** [comment]
@citrus-it do you have any testing notes for this?

**Comment by citrus-it** [comment]
> @citrus-it do you have any testing notes for this?

I do!

Running the updated `cmn-err-1` test with an unpatched gcc produces warnings such as:

```
gcc/testsuite/gcc.dg/format/cmn-err-1.c:33:22: warning: unknown conversion type character 'j' in format [-Wformat=]
gcc/testsuite/gcc.dg/format/cmn-err-1.c:34:22: warning: unknown conversion type character 'z' in format [-Wformat=]
...
FAIL: gcc.dg/format/cmn-err-1.c  (test for excess errors)
```

and running it with the patched version is successful.

Modifying an illumos kernel file to use `%z` (and building with `SHADOW_CCS=` to skip gcc7) produces the same thing:
```
../../oxide/os/ipcc_proto.c:1176:34: error: unknown conversion type character 'z' in format [-Werror=format=]
   cmn_err(CE_NOTE, "SP ALERT %u %zu '%-.*s", action, off,
```

but builds successfully once gcc10 is updated with this patch.

**Comment by rmustacc** [comment]
Thanks, this was landed in e65c791f0cc13ec99962f924a24a192cca739ae2 and tagged as gcc-10.4.0-il-2!

**Pull request closed without merging**